### DISCUSSION TITLE
Laser connection

### DISCUSSION THREAD

**Discussion by jodougie2812** [fra589/grbl-Mega-5X#241] [Q&A]
Hi. I'm using a ramps1.4 board with my Arduino Mega.
I've got 4 axis working using GRBL GRU so that's good.
My question is how do I connect a laser with 3 wires.
Ground  Positive and PWM.. where do these go on the ramps 1.4.
I've tried connecting to D8 but don't have any power showing on D8 D9 or D10

Do I need to alter any settings in GRBL to make it work under PWM
Any advice appreciated

**Comment by fra589** [comment]
Hi @jodougie2812,  
  
Ground and Positive must be connected to any available GND & "+" connection of the RAMPS board:  
https://www.google.com/search?client=firefox-b-d&q=RAMPS+1.4+pinout  
https://github.com/fra589/grbl-Mega-5X/wiki/grbl-Mega-5X-pinout  

Search for "laser pwm" on grbl-Mega-5X / issues / discussions / wiki and you should find some already answered similar questions:  
https://www.google.com/search?client=firefox-b-d&q=laser+pwm+%2Bsite%3Ahttps%3A%2F%2Fgithub.com%2Ffra589%2Fgrbl-Mega-5X
  
If it's not enough, ask a specific question explaining what you did, the result obtained, as well as the expected result.  
Do not forget to specify what is your version of grbl-Mega-5X and as well as the configuration affected by the question.  
  
@++;  
Gauthier.

**Reply by jodougie2812** [reply]
Thanks for your response. I'm now totally confused and my head is about to explode 🤣🤣🤯
When I connect the laser to pos and neg on the ramps board it lights up. I havent connected to the PWM pins yet. Is it the PWM signal which makes the laser turn off until required.
The information about altering configuration h, I think I understand. However the info about altering the settings in cpu maps.h  is confusing.
I'm using the standard,  latest version of GRBL MEGA X on a Arduino Mega wIth a Ramps 1.4 
Thanks in hopeful anticipation

**Reply by jodougie2812** [reply]
These are the settings I made in
config.h

// Chose the spindle pin output :
// SPINDLE_PWM_ON_D8  => 0-12v 16 bits PWM on RAMPS D8
// SPINDLE_PWM_ON_D6  => 0-5v 8bits PWM on RAMPS Servo 2 signal (Mega 2560 D6)
// Uncomment the line which correspond to your hardware
//#define SPINDLE_PWM_ON_D8
#define SPINDLE_PWM_ON_D6

and in cpu_map.h



  // Advanced Configuration Below You should not need to touch these variables

  // Set Timer up to use TIMER4C which is attached to Digital Pin 6 - Ramps Servo 2
  #define SPINDLE_PWM_MAX_VALUE     255.0 // Since it's a 8bit timer
  #ifndef SPINDLE_PWM_MIN_VALUE
  #define SPINDLE_PWM_MIN_VALUE   1   // Must be greater than zero.
  #endif
  #define SPINDLE_PWM_OFF_VALUE     0
  #define SPINDLE_PWM_RANGE         (SPINDLE_PWM_MAX_VALUE-SPINDLE_PWM_MIN_VALUE)

  #define SPINDLE_TCCRA_REGISTER    TCCR4A
  #define SPINDLE_TCCRB_REGISTER    TCCR4B
  #define SPINDLE_OCR_REGISTER      OCR4A
  #define SPINDLE_COMB_BIT          COM4A1

  #define SPINDLE_TCCRA_INIT_MASK (1<<WGM41)
  #define SPINDLE_TCCRB_INIT_MASK ((1<<WGM42) | (1<<WGM43) | (1<<CS41)) 
  #define SPINDLE_OCRA_REGISTER   ICR4 
  #define SPINDLE_OCRA_TOP_VALUE  0xFF // PWM counter reset value. Should be the same as PWM_MAX_VALUE in hex.

  // Define spindle output pins.
  #define SPINDLE_PWM_DDR   DDRH
  #define SPINDLE_PWM_PORT  PORTH
  #define SPINDLE_PWM_BIT   3 // MEGA2560 Digital Pin 6

    #define SPINDLE_PWM_MAX_VALUE     1024.0 // Translates to about 1.9 kHz PWM frequency at 1/8 prescaler
    #ifndef SPINDLE_PWM_MIN_VALUE
      #define SPINDLE_PWM_MIN_VALUE   1   // Must be greater than zero.
    #endif
    #define SPINDLE_PWM_OFF_VALUE     0
    #define SPINDLE_PWM_RANGE         (SPINDLE_PWM_MAX_VALUE-SPINDLE_PWM_MIN_VALUE)

    //Control Digital Pin 8
    #define SPINDLE_TCCRA_REGISTER    TCCR4A
    #define SPINDLE_TCCRB_REGISTER    TCCR4B
    #define SPINDLE_OCR_REGISTER      OCR4C
    #define SPINDLE_COMB_BIT          COM4C1

    // 1/8 Prescaler, 16-bit Fast PWM mode
    #define SPINDLE_TCCRA_INIT_MASK ((1<<WGM40) | (1<<WGM41))
    #define SPINDLE_TCCRB_INIT_MASK ((1<<WGM42) | (1<<WGM43) | (1<<CS41))
    #define SPINDLE_OCRA_REGISTER   OCR4A // 16-bit Fast PWM mode requires top reset value stored here.
    #define SPINDLE_OCRA_TOP_VALUE  0x400 // PWM counter reset value. Should be the same as PWM_MAX_VALUE in hex.

    // Define spindle output pins.
    #define SPINDLE_PWM_DDR   DDRH
    #define SPINDLE_PWM_PORT  PORTH
    #define SPINDLE_PWM_BIT   5 // MEGA2560 Digital Pin 8

  #elif defined (SPINDLE_PWM_ON_D6)
There are a few different descriptions of whats needed and  i'm confused.
Please help
Thank you

**Comment by jodougie2812** [comment]
The laser is connected  to 12v pos and negative on the board. It fires continuously. If I connect PWM to the negative terminal on P6 the laser switches off. Breaking that connection fires the laser.
Running laser GRBL doesn't control the PWM signal on P6
HELP please

**Comment by fra589** [comment]
Hi @jodougie2812,  
  
To change the spindle output from the standard 12V PWM output on D8 to the 5V PWM output on D6, you don't need to change anything in the cpu_map.h file.  
  
You simply have to comment the `//#define SPINDLE_PWM_ON_D8` and uncomment the `#define SPINDLE_PWM_ON_D6` like you did without anything else. The `#if defined(SPINDLE_PWM_ON_D8)` and `#elif defined (SPINDLE_PWM_ON_D6)` preprocessor compilation lines in cpu_map.h will take care of taking the right parameters.
  
Then, the PWM signal on D6 will be drived with standard spindle speed GCode command:
- `S<value> ` with value ~~between 0 to 255~~ _edited_: between "Minimum spindle speed" value ($31) and the "Maximum spindle speed" value ($30).,
- `M3` to start the ~~Spindle~~ Laser,
- `M5` to stop the ~~Spindle~~ Laser.
  
@++;
Gauthier.

**Comment by jodougie2812** [comment]
Good morning
Well. I've taken a fresh download of your software. Changed the code in
configuration.h and reconnected the laser.
The pwm wire when connected to D6 negative turns off the laser. If I break
the connection the laser fires.
However. I'm not getting any control of the laser in my software. I'm using
laser grbl where $32=1
Thanks for your continuing assistance.
John

**Comment by fra589** [comment]
Good morning @jodougie2812,  
  
Can you tell me more about your laser module?... Reference, link to documentation / datasheet, ...  
I don't know a lot of thing about lasergrbl, but I know it use the M5/M3S<value> to drive the power of the laser. So, it should work fine with grbl-Mega-5X.  
I rather suspect a hardware problem.  
Are you able to check the voltage on PWM pin D6 with a multimeter with differents values of M3S<value>?  
  
@++;  
Gauthier.

**Comment by jodougie2812** [comment]
Here you go.  The laser is a cheap Chinese thing supplied with my 3018
machine. It works fine on that machine.
I have put my multimeter across D6 with various settings and there is no
voltage showing on any settings.
My 12 volt supply from the ramps powers up the laser. If I ground the PWM
it switches the laser off. Is that correct?. Should the PWM signal be on
the positive side 🤔

**Reply by fra589** [reply]
> I have put my multimeter across D6 with various settings and there is no voltage showing on any settings.
  
It's not normal. You should have positive voltage on D6, depending of the value of the S parameter.
The S (Spindle speed) parameter value can be between the "Minimum spindle speed" value ($31) and the "Maximum spindle speed" value ($30).  
If you don't change them, by default, $31=550 and $30=12000.  
So:  
- `M3S12000` => will give about 5V on D6,
- `M3S550`     => will give few millivolts on D6,
- `M3S6275`   => will give about 2.5V on D6 (6250 is the middle between 550 and 12000).
  
@++;  
Gauthier.

**Comment by jodougie2812** [comment]
I will try those settings. Thank you

**Comment by jodougie2812** [comment]
Am I connecting the PWM from the laser to D6 positive and the earth to
earth with the 12v positive to 12 volts please 🙏

**Reply by fra589** [reply]
I don't know about the exact laser you have, but I find this with Google:
https://wiki.nejetool.com/doku.php?id=laser_module_308020w
  
This mean that you have to connect 12V positive to 12V POWER, GND to Grbl's GND and SIGNAL to the D6 PWM Grbl's output.  
  
@++;  
Gauthier.

**Comment by jodougie2812** [comment]
Thank you that is most helpful.

**Reply by jodougie2812** [reply]
Well. This is what I did.
I set the values of $30 and $31 as you suggested. I sent some laser g code from both laserGRBL and lightburn I had no voltage at all on D6. So I reflashed the board with the unaltered GRBLMega firmware.  I put my multimeter across D8 pos/neg and had a voltage of 1.2 volts. Running the g code again I had a pulsing voltage between 0 and 1.2 volts. 
I confused as to why the laser activates when powered up and switches off I the earth the PWM wire. I thought to run the pwm wire through the D8 terminals to see if that worked. I didn't for fear of blowing something up.
Any thoughts please 
John

**Comment by fra589** [comment]
I don't suggested any values for $30 & $31... I just tell you what are their default values.
Then, I suggested to use a terminal or software that can send direct Grbl's commands to send the command: `M3S<value of $30>`
If you send this command, you should have the full voltage of PWM.
Have you read the Lasergrbl's documentations about the PWM power laser configuration?
https://lasergrbl.com/configuration/
https://lasergrbl.com/wp-content/uploads/2020/05/Grbl-Configuration-ENG.pdf
Do you what are the S values send by your software when you try to measure the PWM voltage?

**Comment by jodougie2812** [comment]
Ok, I will do that.

**Comment by jodougie2812** [comment]
Wow this is strange....
When I use the fire button in Lightburn with a value of 1000. I get a
voltage of 1.17 volts...
When I increase the value to 1500 I get a voltage of .42 volts
When I increase the value to 2000. I get a negative on the multimeter.
It looks like that if I increase the power in all software (lightburn,
laserGRBL and UGS). The power goes down not up.
Do I need to invert something please.

**Reply by fra589** [reply]
What are your $30 & $31 values ?

**Comment by jodougie2812** [comment]
$30 =12000
$31=500

**Reply by fra589** [reply]
Try the same values in Lightburn with $30=2000 and $31=0

**Comment by jodougie2812** [comment]
😊  Thank you . That has worked..
I now need to discover which pins to use for a probe.
Your help in this community is so very much appreciated

**Reply by fra589** [reply]
> I now need to discover which pins to use for a probe.

Please, read the doc: https://github.com/fra589/grbl-Mega-5X/wiki
More precisely: https://github.com/fra589/grbl-Mega-5X/wiki/grbl-Mega-5X-pinout
  
@++;  
Gauthier

**Comment by jodougie2812** [comment]
I spoke too soon. I suspect either a dodgy ramps or laser?...I am replacing
both
Thankyou

**Comment by jodougie2812** [comment]
Thank you.

**Reply by jodougie2812** [reply]
Now got my replacement boards. So I hope I can report success tomorrow

**Reply by jodougie2812** [reply]
Sorry to be on here again. I'm very puzzled. I've put in a new Arduino Mega and Ramps 1.4 but I still have laser problems. I've read so much stuff on what to do. Much of it is beyond my limited knowledge. I'm too old to learn. Or it is contradictory.
Here's what I've got now.....
I've flashed the arduino with Grbl Mega 5x with the modified configuration.h file to use D6 as the PWM for my laser..
Running a simple file from LaserGRBL and Lightburn I get the same result. I have a modulating voltage on D8 and a static 5v on D6, no modulation.. any ideas as to what I'm doing wrong.....I'm about to give up and just use the laser on my 3 axis CNC but I really want to use it on the rotary A axis

**Reply by jodougie2812** [reply]
If I connect the PWM lead to positive or negative on D8 I get the laser firing but not modulating with the PWM. It works well with my CNC shield

**Reply by jodougie2812** [reply]
Essentially.  What I don't understand is why the output from the 3018 board and the pwm pin on the CNC shield works. Whereas the PWM output from the RAMPS shield doesnt

**Reply by fra589** [reply]
Hi @jodougie2812 

Can you share your full edited version of grbl-Mega-5X (in a .zip file)?  
Please, also post your full config (result of $$ and $I) grbl's commands. 
  
I'm not at home, but I have some moments where I can look at it and try to understand what happen to you.

@++;
Gauthier.
